New event system

Spoffy opened the pull request.

Spoffy committed:
Adds local version of new event system

Spoffy committed:
Adds core networked event code, without edge cases

Spoffy committed:
Removes separate multi-machine event handling

Spoffy committed:
Removes event system client array reservations

Spoffy committed:
Reinstates global event forwarding on the server

Spoffy committed:
Tidies up serverside event system naming

Spoffy committed:
Tidies clientside event system and fixes object topics not working

Spoffy committed:
Adds WIP TODOs to fn_event_onPlayerDisconnected

Spoffy committed:
Fixes errors in event system and adds healthCheck function

Spoffy committed:
Adds potentially working unsubscribe to event system

Spoffy committed:
Adds disconnect handling to event system

veteran29 commented on paradigm/global/functions/events/fn_event_subscribe.sqf:
```diff
@@ -0,0 +1,96 @@
+        {
+               _x isEqualTo 0
+            || !(_x isEqualTo clientOwner)
+            || isServer && !(_x isEqualTo 2)
```
```suggestion
            || (_x isNotEqualTo clientOwner)
            || isServer && (_x isNotEqualTo 2)
```

veteran29 commented on paradigm/global/functions/events/fn_event_subscribe.sqf:
```diff
@@ -0,0 +1,96 @@
+    _event = [_event, hashValue ""];
+} else {
+    _event = [_event # 0, hashValue (_event # 1)]
+};
```
I think we should hash topics only when they're not a string. Also the default "global" topic should be an empty string, not a hash of one.
```suggestion
if (_event isNotEqualType []) then {
    _event = [_event, ""];
} else {
    if (_event isEqualTypeParams [nil, ""]) exitWith {};
    _event = [_event # 0, hashValue (_event # 1)];
};
```

Spoffy replied:
Should be done in the latest push

veteran29 commented on paradigm/global/functions/events/fn_event_subscribe.sqf:
```diff
@@ -0,0 +1,96 @@
+    _event = [_event # 0, hashValue (_event # 1)]
+};
+
+_event params ["_eventName", "_topic"];
```
We should validate that event name is a string.
```suggestion
_event params [["_eventName", nil, [""]], "_topic"];
```

Spoffy replied:
Added :)

veteran29 commented on paradigm/server/functions/events/fn_event_forward.sqf:
```diff
@@ -0,0 +1,48 @@
+
+private _machinesListeningToThisOrigin = _forwardingForOriginMachineId
+    getOrDefault [_originMachineId, createHashMap]
+    getOrDefault [_eventHash, []];
```
No need to hash event array as we already will have only hash-key "legal" elements in array (strings).
```suggestion
private _forwardingForOriginMachineId = localNamespace getVariable "para_event_forwardingForOriginMachineId";

private _machinesListeningToAllOrigins = _forwardingForOriginMachineId
    getOrDefault [0, createHashMap]
    getOrDefault [_event, []];

private _machinesListeningToThisOrigin = _forwardingForOriginMachineId
    getOrDefault [_originMachineId, createHashMap]
    getOrDefault [_event, []];
```

Spoffy replied:
All done. Needed an ever so slightly different implementation though so that the client receives the original topic in its callback. (It didn't in the original version either, it was a bug)

veteran29 commented on paradigm/server/functions/events/fn_event_startForwardingMatchingEvents.sqf:
```diff
@@ -0,0 +1,65 @@
+{
+    private _listeningMachineIds = _forwardingForOriginMachineId
+        getOrDefault [_x, createHashMap, true]
+        getOrDefault [_eventHash, [], true];
```
No need to hash event array as we already will have only hash-key "legal" elements in array (strings).
```suggestion
private _listenerMachineIdReference = _machineIdReferences get remoteExecutedOwner;

// Register the client sending the request as wanting events forwarded from _clients
{
    private _listeningMachineIds = _forwardingForOriginMachineId
        getOrDefault [_x, createHashMap, true]
        getOrDefault [_event, [], true];
```

Spoffy replied:
Done!

veteran29 commented on paradigm/server/functions/events/fn_event_stopForwardingMatchingEvents.sqf:
```diff
@@ -0,0 +1,56 @@
+
+{
+    private _originMachineId = _x;
+    private _originMachineForwarding = _forwardingForOriginMachineId getOrDefault [_originMachineId, createHashMap] getOrDefault [_eventHash, []];
```
No need to hash event array as we already will have only hash-key "legal" elements in array (strings).
```suggestion
private _machineIdReferences = localNamespace getVariable "para_event_machineIdReferences";
private _forwardingForOriginMachineId = localNamespace getVariable "para_event_forwardingForOriginMachineId";
private _listenerMachineIdReference = _machineIdReferences get remoteExecutedOwner;

private _globalForwarding = _forwardingForOriginMachineId getOrDefault [0, createHashMap] getOrDefault [_event, []];

private _clientsToStopForwardingEvent = [];

{
    private _originMachineId = _x;
    private _originMachineForwarding = _forwardingForOriginMachineId getOrDefault [_originMachineId, createHashMap] getOrDefault [_event, []];
```

Spoffy replied:
Done!

veteran29 commented on paradigm/global/functions/events/fn_event_startForwardingMatchingEventsToServer.sqf:
```diff
@@ -0,0 +1,23 @@
+
+params ["_event"];
+
+localNamespace getVariable "para_event_eventsToForward" set [hashValue _event, true];
```
See suggestion to `subscribe`.
```suggestion
localNamespace getVariable "para_event_eventsToForward" set [_event, true];
```

Spoffy replied:
Done!

veteran29 commented on paradigm/global/functions/events/fn_event_stopForwardingMatchingEventsToServer.sqf:
```diff
@@ -0,0 +1,26 @@
+private _eventsToforward = localNamespace getVariable "para_event_eventsToForward";
+private _eventHash = hashValue _event;
+
+_eventsToForward deleteAt _eventHash;
```
See suggestion to `subscribe`.
```suggestion
_eventsToForward deleteAt _event;
```

Spoffy replied:
Done!

veteran29 commented on paradigm/global/functions/events/fn_event_callRegisteredHandlers.sqf:
```diff
@@ -0,0 +1,86 @@
+
+private _handlerIdsToCall = [];
+
+private _globalTopic = hashValue "";
```
See suggestion to `subscribe`.
```suggestion
private _globalTopic = "";
```

Spoffy replied:
Done!

Spoffy committed:
Fixes various bugs in the event system

- Fixes players not being initialised if they're already connected when system is initialised
- Fixes global listening not working due to the handler being removed from the cache
- Fixes handlers not being removed from cache when unsubscribed.
- Fixes bad hashing on event trigger.

Spoffy committed:
Adds event system unit tests

Spoffy committed:
Fixes global event handlers not being called when triggered remotely

Spoffy committed:
Fixes server forwarding events to the originating client

Spoffy committed:
Fixes dedicated server not being initialised as a client in event system

Spoffy committed:
Makes event handlers receive raw topic instead of hash

Spoffy committed:
Fixes topic-specific events not being called

Spoffy committed:
Switches to isNotEqualTo in fn_event_subscribe.sqf

Co-authored-by: Filip Maciejewski <veteran29@users.noreply.github.com>

Spoffy committed:
Fixes event system double handler calls and no global event triggering

Spoffy committed:
Fixes bad general event hash when forwarding

Spoffy committed:
Changes event hashes to preserve string topics

Spoffy committed:
Updates event health check with new hashing

Spoffy committed:
Adds validation to fn_event_subscribe that the eventName is a string

Spoffy committed:
Misc. fixes to event system + event system tests

Spoffy committed:
Fixes event system unsubscribe, and race conditions on joining.

Spoffy committed:
Updates health check function to work with recent changes

veteran29 commented on paradigm/global/functions/events/fn_event_callRegisteredHandlers.sqf:
```diff
@@ -0,0 +1,65 @@
+    getOrDefault [_originMachineId, createHashMap]
+    getOrDefault [_eventName, createHashMap];
+
+_handlerIdsToCall = _handlerIdsToCall + (_machineSpecificEventListenersByTopic getOrDefault [_globalTopic, []]);
```
No need to always merge with empty array. 
```suggestion
private _handlerIdsToCall = _machineSpecificEventListenersByTopic getOrDefault [_globalTopic, []];
```

veteran29 commented on paradigm/global/functions/events/fn_event_callRegisteredHandlers.sqf:
```diff
@@ -0,0 +1,65 @@
+
+private _eventListenersByOrigin = localNamespace getVariable "para_event_listenersByEventOrigin";
+
+private _handlerIdsToCall = [];
```
See below.
```suggestion
```

veteran29 replied:
Actually see above as GitHub added these in opposite order ;D

veteran29 commented on paradigm/global/functions/events/fn_event_convertEventToHashableEvent.sqf:
```diff
@@ -0,0 +1,28 @@
+params [["_event", nil, [[]], 2]];
+
+if (_event # 1 isEqualType "") exitWith {
+    _event + []
```
What's the purpose of this? Did you want to return copy of `_event` array? If so `+_event` is enough.

Spoffy replied:
`+_event` is a deep copy, `_event + []` is a shallow copy, which is what I wanted here :) Though given both should be strings, it doesn't really matter either way now!

veteran29 commented on paradigm/global/functions/events/fn_event_startForwardingMatchingEventsToServer.sqf:
```diff
@@ -0,0 +1,34 @@
+
+{
+    private _hashableEvent = _x;
+    _eventsToForward set [_hashableEvent, true];
```
IMO it's readable enough without the additional variable.
```suggestion
    _eventsToForward set [_x, true];
```

veteran29 commented on paradigm/global/functions/events/fn_event_stopForwardingMatchingEventsToServer.sqf:
```diff
@@ -0,0 +1,34 @@
+
+{
+    private _hashableEvent = _x;
+    _eventsToForward deleteAt _hashableEvent;
```
Also IMO readable enough without the addional variable.
```suggestion
    _eventsToForward deleteAt _x;
```

veteran29 commented on paradigm/global/functions/events/fn_event_subscribe.sqf:
```diff
@@ -0,0 +1,99 @@
+
+params [["_clients", [clientOwner]], "_event", "_handler"];
+
+// Standardise event format, and hash topic to make sure it's a string.
```
```suggestion
// Standardise event format
```

veteran29 commented on paradigm/global/functions/events/fn_event_subscribe.sqf:
```diff
@@ -0,0 +1,99 @@
+
+
+
+
```
Excess whitespace
```suggestion
```

veteran29 reviewed (approved):
Mostly looks good to me.

A lot of functions have placeholder data in their headers, please fill them.

Spoffy committed:
Removes empty array merge from fn_event_callRegisteredHandlers

Co-authored-by: Filip Maciejewski <veteran29@users.noreply.github.com>

Spoffy committed:
Updates fn_event_startForwardingMatchingEventsToServer to remove spare variable

Co-authored-by: Filip Maciejewski <veteran29@users.noreply.github.com>

Spoffy committed:
Removes un-needed variable from fn_event_stopForwardingMatchingEventsToServer

Co-authored-by: Filip Maciejewski <veteran29@users.noreply.github.com>

Spoffy committed:
Updates comments and whitespace for event system

Co-authored-by: Filip Maciejewski <veteran29@users.noreply.github.com>
